File: 0912-sort-an-array/0912-sort-an-array.java
class Solution {
    void swap(int pivot, int ind, int[] nums) {
        int t = nums[pivot];
        nums[pivot] = nums[ind];
        nums[ind] = t;
    }
    void sort(int[] nums, int left, int right) {
        if(left >= right) return;
        int pivot = (int)(Math.random()*(right-left+1))+left;
        int piv = nums[pivot];
        swap(pivot, right, nums);
        int l = left;
        int r = right-1;
        while(l<=r) {
            if(nums[l] <= piv) l++;
            else {
                swap(l, r, nums);
                r--;
            }
        }
        swap(l, right, nums);
        sort(nums, left, l-1);
        sort(nums, l+1, right);
    }
    public int[] sortArray(int[] nums) {
        sort(nums, 0, nums.length-1);
        return nums;
    }
}

File: 0912-sort-an-array/0912-sort-an-array.cpp
class Solution {
public:
    void merge(vector<int>& nums, vector<int>& aux, int left, int mid, int right) {
        for(int i=left; i<=right; i++) aux[i] = nums[i];
        int i = left;
        int l = left;
        int r = mid;
        left = mid+1;
        while(l <= r && left <= right) {
            if(aux[l] < aux[left]) {
                nums[i++] = aux[l++];
            }
            else {
                nums[i++] = aux[left++];
            }
        }
        if(l <= r) {
            while(l <= r) nums[i++] = aux[l++];
        }
        if(left <= right) {
            while(left <= right) nums[i++] = aux[left++];
        }
    }
    void sort(vector<int>& nums, vector<int>& aux, int left, int right) {
        if(left >= right) return;
        int mid = left + (right - left) / 2;
        sort(nums, aux, left, mid);
        sort(nums, aux, mid+1, right);
        merge(nums, aux, left, mid, right);
    }
    vector<int> sortArray(vector<int>& nums) {
        vector<int> aux(nums.size(), 0);
        sort(nums, aux, 0, nums.size()-1);
        return nums;
    }
};